### PR TITLE
Refactor time utilities

### LINKED ISSUE #1241: Infraction listing for permanent infractions lists "expired" time remaining
A permanent infraction will have its "Remaining" attribute displayed as "Expired" due to the following lines of code:

https://github.com/python-discord/bot/blob/5df2fa2f948852826322409103c91565390110fe/bot/exts/moderation/infraction/management.py#L264

Remaining is "None" not because it has expired, but because there is no defined time remaining for an infraction that is not scheduled to expire. Previously, there was a suggestion in issue #668:

> For fun, we could put (∞) next to permanent infractions.

I think this could be a good choice here. It's also unlikely to introduce much complexity, since all that's needed is an additional case to the "time remaining" if statement to handle unexpired permanent infractions.

### PR DESCRIPTION
**Pull request by MarkKoz**
Fixes #1241 (in d8e237b)

The implementations had some redundancies. Furthermore, the functions were commonly being used with either `datetime`s or ISO 8601 strings, but some functions only accepted one of them rather than both types. This resulted in ugly call sites because callers had to perform conversions themselves. Lastly, some functions were more tightly coupled (mostly naming-wise since the switch to Discord timestamps) to infractions than they needed to be.

The changes are described well by the commit history. Some highlights for your convenience:

* All functions which took a time now support both ISO 8601 and `datetime`s. In fact, they support much more - almost any type supported by `arrow.get()`
* `humanize_delta` supports a `relativedelta`, kwargs to forward to a `relativedelta`, or 2 timestamps (any type described above) to forward to `relativedelta`.

I suggest reviewing this by going through each commit rather than looking at the entire diff. A lot of the diff is caused by renaming and changing the way the time functions are imported. Fortunately for you, these changes are in separate commits from the things that really matter (non-trivial changes).

### CODE REVIEW

File: bot/utils/time.py [hunk 1]
```diff
@@ -60,33 +69,138 @@ def _stringify_time_unit(value: int, unit: str) -> str:
...
+    Format a timestamp as a Discord-flavored Markdown timestamp.
+
+    `timestamp` can be any type supported by the single-arg `arrow.get()`, except for a `tzinfo`.
+    """
+    timestamp = int(arrow.get(timestamp).timestamp())
+    return f"<t:{timestamp}:{format.value}>"
+
+
+@overload
+def humanize_delta(
```
Comment (Numerlor) [inline bot/utils/time.py:83]: A region to collapse the overloads would be useful

File: bot/utils/time.py [hunk 2]
```diff
@@ -161,71 +267,58 @@ def relativedelta_to_timedelta(delta: relativedelta) -> datetime.timedelta:
...
+    """
+    Format `timestamp` as a relative Discord timestamp.
 
+    A relative timestamp describes how much time has elapsed since `timestamp` or how much time
+    remains until `timestamp` is reached.
 
-def format_infraction(timestamp: str) -> str:
-    """Format an infraction timestamp to a discord timestamp."""
-    return discord_timestamp(dateutil.parser.isoparse(timestamp))
+    `timestamp` can be any type supported by the single-arg `arrow.get()`, except for a `tzinfo`.
```
Comment (Numerlor) [inline bot/utils/time.py:277]: I think here (and in other docstrings) the param typehint alone should be enough without further explaining it in the docstring; if necessary the comment above the type alias explains where exactly the types come from

Comment (MarkKoz) [reply bot/utils/time.py:277]: I don't know... I think it doesn't hurt to keep these docstrings just to be safe on the clarity of what it is.

Comment (MarkKoz) [reply bot/utils/time.py:277]: At least they'd have enough information to look up the arrow docs.

File: bot/utils/time.py (hunk 1 above)
Comment (MarkKoz) [reply bot/utils/time.py:84]: Done ef09481

File: bot/exts/moderation/infraction/_scheduler.py [hunk 3]
```diff
@@ -363,20 +363,15 @@ async def deactivate_infraction(
...
-        expiry = dateutil.parser.isoparse(expiry).replace(tzinfo=None) if expiry else None
-        created = time.format_infraction_with_duration(inserted_at, expiry)
-
         log_content = None
         log_text = {
             "Member": f"<@{user_id}>",
             "Actor": f"<@{actor}>",
             "Reason": infraction["reason"],
-            "Created": created,
+            "Created": time.format_with_duration(infraction["inserted_at"], infraction["expires_at"]),
```
Comment (ChrisLovering) [inline bot/exts/moderation/infraction/_scheduler.py:374]: When manually unbanning someone with a perma ban, this raises the below error. I assume this is due to expires_at being `None`
```
bot_1       | Traceback (most recent call last):
bot_1       |   File "/usr/local/lib/python3.9/site-packages/discord/ext/commands/core.py", line 85, in wrapped
bot_1       |     ret = await coro(*args, **kwargs)
bot_1       |   File "/bot/bot/exts/moderation/infraction/infractions.py", line 270, in unban
bot_1       |     await self.pardon_infraction(ctx, "ban", user)
bot_1       |   File "/bot/bot/exts/moderation/infraction/_scheduler.py", line 292, in pardon_infraction
bot_1       |     log_text = await self.deactivate_infraction(response[0], send_log=False, notify=notify)
bot_1       |   File "/bot/bot/exts/moderation/infraction/_scheduler.py", line 374, in deactivate_infraction
bot_1       |     "Created": time.format_with_duration(infraction["inserted_at"], infraction["expires_at"]),
bot_1       |   File "/bot/bot/utils/time.py", line 303, in format_with_duration
bot_1       |     duration = humanize_delta(timestamp, other_timestamp, max_units=max_units)
bot_1       |   File "/bot/bot/utils/time.py", line 196, in humanize_delta
bot_1       |     start = arrow.get(args[1]) if len(args) == 2 else arrow.utcnow()
bot_1       |   File "/usr/local/lib/python3.9/site-packages/arrow/api.py", line 80, in get
bot_1       |     return _factory.get(*args, **kwargs)
bot_1       |   File "/usr/local/lib/python3.9/site-packages/arrow/factory.py", line 228, in get
bot_1       |     raise TypeError("Cannot parse argument of type None.")
bot_1       | TypeError: Cannot parse argument of type None.
```

File: bot/exts/moderation/infraction/management.py [hunk 4]
```diff
@@ -167,8 +163,8 @@ async def infraction_edit(
                 self.infractions_cog.schedule_expiration(new_infraction)
 
             log_text += f"""
-                Previous expiry: {until_expiration(infraction['expires_at']) or "Permanent"}
-                New expiry: {until_expiration(new_infraction['expires_at']) or "Permanent"}
+                Previous expiry: {time.until_expiration(infraction['expires_at'])}
```
Comment (ChrisLovering) [inline bot/exts/moderation/infraction/management.py:166]: Similar as the above, this fails with the below error when trying to edit an infraction without an expiry.
```
bot_1       | Traceback (most recent call last):
bot_1       |   File "/usr/local/lib/python3.9/site-packages/discord/ext/commands/core.py", line 85, in wrapped
bot_1       |     ret = await coro(*args, **kwargs)
bot_1       |   File "/bot/bot/exts/moderation/infraction/management.py", line 134, in infraction_edit
bot_1       |     expiry = time.format_with_duration(duration)
bot_1       |   File "/bot/bot/utils/time.py", line 303, in format_with_duration
bot_1       |     duration = humanize_delta(timestamp, other_timestamp, max_units=max_units)
bot_1       |   File "/bot/bot/utils/time.py", line 196, in humanize_delta
bot_1       |     start = arrow.get(args[1]) if len(args) == 2 else arrow.utcnow()
bot_1       |   File "/usr/local/lib/python3.9/site-packages/arrow/api.py", line 80, in get
bot_1       |     return _factory.get(*args, **kwargs)
bot_1       |   File "/usr/local/lib/python3.9/site-packages/arrow/factory.py", line 228, in get
bot_1       |     raise TypeError("Cannot parse argument of type None.")
bot_1       | TypeError: Cannot parse argument of type None
```

File: bot/exts/moderation/infraction/_scheduler.py [hunk 5]
```diff
@@ -132,7 +132,7 @@ async def apply_infraction(
         infr_type = infraction["type"]
         icon = _utils.INFRACTION_ICONS[infr_type][0]
         reason = infraction["reason"]
-        expiry = time.format_infraction_with_duration(infraction["expires_at"])
+        expiry = time.format_with_duration(infraction["expires_at"])
```
Comment (ChrisLovering) [inline bot/exts/moderation/infraction/_scheduler.py:135]: I'm not entirely sure why, or if its the same issue as the other, but this is erroring even when I give an expiry date 
```
bot_1       | 2021-09-02 20:14:11 | bot.exts.backend.error_handler | ERROR | Error executing command invoked by Chrisjl#2655: &mute 196664644113268736 7d
bot_1       | Traceback (most recent call last):
bot_1       |   File "/usr/local/lib/python3.9/site-packages/discord/ext/commands/core.py", line 85, in wrapped
bot_1       |     ret = await coro(*args, **kwargs)
bot_1       |   File "/bot/bot/exts/moderation/infraction/infractions.py", line 159, in tempmute
bot_1       |     await self.apply_mute(ctx, user, reason, expires_at=duration)
bot_1       |   File "/bot/bot/exts/moderation/infraction/infractions.py", line 312, in apply_mute
bot_1       |     await self.apply_infraction(ctx, infraction, user, action())
bot_1       |   File "/bot/bot/exts/moderation/infraction/_scheduler.py", line 135, in apply_infraction
bot_1       |     expiry = time.format_with_duration(infraction["expires_at"])
bot_1       |   File "/bot/bot/utils/time.py", line 303, in format_with_duration
bot_1       |     duration = humanize_delta(timestamp, other_timestamp, max_units=max_units)
bot_1       |   File "/bot/bot/utils/time.py", line 196, in humanize_delta
bot_1       |     start = arrow.get(args[1]) if len(args) == 2 else arrow.utcnow()
bot_1       |   File "/usr/local/lib/python3.9/site-packages/arrow/api.py", line 80, in get
bot_1       |     return _factory.get(*args, **kwargs)
bot_1       |   File "/usr/local/lib/python3.9/site-packages/arrow/factory.py", line 228, in get
bot_1       |     raise TypeError("Cannot parse argument of type None.")
bot_1       | TypeError: Cannot parse argument of type None.
bot_1       | /usr/local/lib/python3.9/asyncio/events.py:80: RuntimeWarning: coroutine 'Infractions.apply_mute.<locals>.action' was never awaited
bot_1       |   self._context.run(self._callback, *self._args)
bot_1       | RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

File: bot/exts/moderation/stream.py [hunk 6]
```diff
@@ -134,10 +134,10 @@ async def stream(self, ctx: commands.Context, member: discord.Member, duration:
 
         await member.add_roles(discord.Object(Roles.video), reason="Temporary streaming access granted")
 
-        await ctx.send(f"{Emojis.check_mark} {member.mention} can now stream until {discord_timestamp(duration)}.")
+        await ctx.send(f"{Emojis.check_mark} {member.mention} can now stream until {time.discord_timestamp(duration)}.")
 
         # Convert here for nicer logging
-        revoke_time = format_infraction_with_duration(str(duration))
+        revoke_time = time.format_with_duration(duration)
```
Comment (ChrisLovering) [inline bot/exts/moderation/stream.py:140]: This errors with the below log
```
bot_1       | 2021-09-02 20:18:35 | bot.exts.backend.error_handler | ERROR | Error executing command invoked by Chrisjl#2655: &stream 126811506632294400
bot_1       | Traceback (most recent call last):
bot_1       |   File "/usr/local/lib/python3.9/site-packages/discord/ext/commands/core.py", line 85, in wrapped
bot_1       |     ret = await coro(*args, **kwargs)
bot_1       |   File "/bot/bot/exts/moderation/stream.py", line 140, in stream
bot_1       |     revoke_time = time.format_with_duration(duration)
bot_1       |   File "/bot/bot/utils/time.py", line 303, in format_with_duration
bot_1       |     duration = humanize_delta(timestamp, other_timestamp, max_units=max_units)
bot_1       |   File "/bot/bot/utils/time.py", line 196, in humanize_delta
bot_1       |     start = arrow.get(args[1]) if len(args) == 2 else arrow.utcnow()
bot_1       |   File "/usr/local/lib/python3.9/site-packages/arrow/api.py", line 80, in get
bot_1       |     return _factory.get(*args, **kwargs)
bot_1       |   File "/usr/local/lib/python3.9/site-packages/arrow/factory.py", line 228, in get
bot_1       |     raise TypeError("Cannot parse argument of type None.")
bot_1       | TypeError: Cannot parse argument of type None.
```

**Review by ChrisLovering — CHANGES_REQUESTED** [review]
Got a few errors while testing this out in the test server they all *seem* to be caused by the same thing though

File: bot/exts/moderation/infraction/management.py (hunk 4 above)
Comment (ChrisLovering) [reply bot/exts/moderation/infraction/management.py:166]: Sorry, this is comment is on the wrong line. This error is actually for line 123 in here.

**Comment by MarkKoz** [comment]
Looks like all the errors are with `humanize_delta` not understanding `None` as the second timestamp. If only one timestamp is passed, then it's relative to now, but it does not treat a `None` second timestamp as equivalent to not having passed it at all. I suppose the best fix is to allow it to be None; having to check for None outside seems too annoying.

**Comment by MarkKoz** [comment]
I left `humanize_delta` alone since all calls already check for None and when it is None they don't want to use the current time actually. I just changed `format_duration` to pass the current time if the 2nd arg is None.

**Comment by MarkKoz** [comment]
> 
> 
> Hey @MarkKoz 😅 We merged #1815, which migrated the bot to disocrd.py 2.0a0 and made aaallll the timestamp tz aware, probably breaking the whole PR 😬

That sucks. This PR has been ready before that one was even open. I'm not going to be touching this PR until people commit to reviewing it; I'd rather keep the merge commits to a minimum.

**Comment by MarkKoz** [comment]
Updated the PR with changes from main since @HassanAbouelela kindly expressed interest in reviewing this. I'd still like to test things by running the bot (which the reviewer will hopefully be doing regardless). I did do that originally, but it'd be good to do it again since the "merge" operation was complicated. Anyway, will get around to testing it myself soon.

**Review by wookie184 — APPROVED** [review]
Read through the commits and all changes seem to make sense, and have tested the things I can think of that this could affect. Everything seems to be working well.

**Review by HassanAbouelela — APPROVED** [review]
This is a huge PR, and it improves our codebase significantly. Thanks for putting in all this work. Everything looks good to me.

**Pull request merged**